### PR TITLE
Initial push of symmetric cipher - AES

### PR DESCRIPTION
**Pull request by saru95**
@panzone I've pushed this as an initial draft of AES. Right now the KDF is in the same file. I'll separate it after the issues for AES (if any) are fixed. 

Code to test:

```chpl
use Crypto;
var a = new AES(256, "cbc");
var salt = new CryptoBuffer("random_salt");
var k = a.generateKey("random_key", salt);
writeln("Generated Key: ", k.toHex());
var iv = a.generateIV();
writeln("Generated IV: ", iv.toHex());

var msg = new CryptoBuffer("foo_bar");
writeln("Original Message: ", msg.toHex());
var ct = a.encrypt(msg, k, iv);
writeln("Obtained Ciphertext: ", ct.toHex());

var orig = a.decrypt(ct, k, iv);
writeln("Obtaineds Plaintext: ", orig.toHex());
```

### CODE REVIEW

File: Crypto.chpl [hunk 1]
```diff
@@ -47,4 +50,70 @@ module Crypto {
...
   }
 
+
+  /* AES Symmetric cipher */
+  class AES {
+    var cipher: EVP_CIPHER_PTR;
+    var bitLen: int;
+
+    proc AES(bits: int, mode: string) {
+      if (bits == 128 && mode == "cbc") {
```
Comment (panzone) [inline Crypto.chpl]: Right now I would only do 128/196/256 CBC mode.

File: CryptoSupport/aesSupport.chpl [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                                outx: c_ptr(c_uchar)): c_int;
+
+  proc getPBKDFKey(userKey: string, bitLen: int) {
+    var key: [0..(bitLen-1)] uint(8);
+    var salt: [0..(bitLen-1)] uint(8);
+    var userKeyLen = userKey.length;
+    var iterCount = 1000;
+    const md = EVP_sha256();
+
+    RAND_bytes(c_ptrTo(salt): c_ptr(c_uchar), bitLen: c_int);
```
Comment (panzone) [inline CryptoSupport/aesSupport.chpl]: This cannot work: the salt must be provided by the user. If you generate here a random salt, the user will not be able to generate the correct key even if they use the correct password.

File: CryptoSupport/aesSupport.chpl [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    EVP_CIPHER_CTX_init(ctx);
+
+    /* Get buffer contents */
+    var keyData = key.getBuffData();
+    var ivData = IV.getBuffData();
+    var plaintextData = plaintext.getBuffData();
+    var plaintextLen = plaintext.getBuffSize();
+
+    /* Allocating space for obtaining the ciphertext */
+    var ciphertextLen = plaintextLen + 16; // 16 is the MAX_BLOCK_SIZE for AES
```
Comment (panzone) [inline CryptoSupport/aesSupport.chpl]: 16 is the MAX_BLOCK_SIZE only for AES128.

Comment (saru95) [inline CryptoSupport/aesSupport.chpl]: Okay. I'll increase it. 👍

File: Crypto.chpl (hunk 1 above)
Comment (saru95) [inline Crypto.chpl]: What's the reason behind this?

File: CryptoSupport/aesSupport.chpl (hunk 2 above)
Comment (saru95) [inline CryptoSupport/aesSupport.chpl]: Cool. I'll write a separate routine for generating the salt.

File: Crypto.chpl (hunk 1 above)
Comment (panzone) [inline Crypto.chpl]: Reducing complexity in a single PR. Every combination of bit length/mode will need careful testing, so it's better to do one thing at the time.

File: CryptoSupport/aesSupport.chpl (hunk 2 above)
Comment (saru95) [inline CryptoSupport/aesSupport.chpl]: Does the salt go in with the user specified key like 

```chpl
var key = generateKey("random_key", salt)
```
where salt is also a buffer?

Comment (panzone) [inline CryptoSupport/aesSupport.chpl]: Yes.

File: CryptoSupport/aesSupport.chpl (hunk 3 above)
Comment (saru95) [inline CryptoSupport/aesSupport.chpl]: Wikipedia says block size for all AES versions is 128 bits (16 bytes). 
https://en.wikipedia.org/wiki/Advanced_Encryption_Standard

**Pull request merged**